### PR TITLE
Make MosaicGPT a HuggingFace PreTrainedModel

### PR DESCRIPTION
**Pull request by dakinggg**
(you can ignore the long list of commits this PR claims to have, it is just because I started working off of the 0.13.2 merge PR before it was merged to main)

This PR makes MosaicGPT a subclass of transformers.PreTrainedModel, allowing us to call `generate` and other HF utils. Note: there are still some todos, like supporting kv caching, but I think all of those should be called out with explicit errors in the code, and should be easy to add once this base is in. Note: this PR will wait for the KV caching PR to be merged to main before merging.

Testing todos:
- [x] make sure generate works properly on a trained model, both with and without padding
Shows 7b with batched (i.e. padded) and not batched output (i.e. not padded)
```
['I believe that MosaicML is the best company in the world. Someday, I hope to be a part of it.\nI am a senior at the University of Texas at', '<|endoftext|><|endoftext|>Q: What city is the capital of the United States?\nA: Washington, D.C. The capital of the United States is Washington, D.C.\n', '<|endoftext|><|endoftext|><|endoftext|>Q: What is the capital of the United States?\nA: Washington, D.C. The capital of the United States is Washington, D.C.\n'] 2023-03-20 22:53:23

I believe that MosaicML is the best company in the world. Someday, I hope to be a part of it.
I am a senior at the University of Michigan,

Q: What city is the capital of the United States?
A: Washington, D.C. The capital of the United States is Washington, D.C.

Q: What is the capital of the United States?
A: Washington, D.C. The capital of the United States is Washington, D.C.
```
- [x] compare training run before and after, need to test with all three attention implementations
(base run is from before this PR and uses flash attention)
Screen Shot 2023-03-20 at 4 07 59 PM

- [x] add tests for save_pretrained and from_pretrained
- [x] add kv caching support after vitaliy's pr

### CODE REVIEW

**Comment by dakinggg** [comment]
test generate script for the curious: https://gist.github.com/dakinggg/b960c934cf021f92b6e909b12b8513a1

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (vchiley) [inline examples/llm/src/models/layers/gpt_blocks.py:51]: isn't this already checked in MosaicGPTConfig?

Comment (vchiley) [reply examples/llm/src/models/layers/gpt_blocks.py:51]: but also [hypocrisy](https://github.com/mosaicml/examples/pull/244#discussion_r1142698694)

File: examples/llm/src/models/layers/attention.py
Comment (vchiley) [inline examples/llm/src/models/layers/attention.py:236]: love the confidence in setting triton as default ❤️

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dakinggg) [reply examples/llm/src/models/layers/gpt_blocks.py:51]: haha woops, i missed this one. yeah should be fine to check in the config as i do

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:34]: I thought we were going to start moving away from this name... (cuz OpenAI has a copyright on the gpt name) 
@abhi-mosaic how does 'mosaic_llm' sound?
(this may be outside the scope of this PR)

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:136]: why change it from `key_padding_mask`?

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:165]: I think once you merge my pr, this should be taken care of.

File: examples/llm/src/models/layers/gpt_blocks.py
Comment (dakinggg) [reply examples/llm/src/models/layers/gpt_blocks.py:47]: Ah, this was left over from when I needed to pass the kwargs along to the unknown causal attn class. will del them here, and remove from the `MultiheadAttention` constructor call.

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:136]: yeah, huggingface expects it to be called `attention_mask`. In principle, we could change its name in `prepare_inputs_for_generation`, but i'd rather just follow their expected forward signature.

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:165]: I agree.

File: examples/resnet_imagenet/main.py
Comment (vchiley) [inline examples/resnet_imagenet/main.py:23]: ?

Comment (dakinggg) [reply examples/resnet_imagenet/main.py:23]: i have no idea how this slipped through previously, but it was causing my pyright errors or something. all package imports should be fully qualified.

File: examples/llm/tests/test_model.py
Comment (dakinggg) [reply examples/llm/tests/test_model.py:370]: fair enough

File: examples/llm/tests/test_model.py
Comment (vchiley) [inline examples/llm/tests/test_model.py:382]: add `('torch', 'gpu')`,

File: examples/llm/tests/test_model.py
Comment (vchiley) [inline examples/llm/tests/test_model.py:549]: Why not also check if parameters are the same? (torch.equal)

File: examples/llm/tests/test_model.py
Comment (vchiley) [inline examples/llm/tests/test_model.py:477]: add `('torch', 'gpu'),`

File: examples/llm/tests/test_model.py
Comment (dakinggg) [reply examples/llm/tests/test_model.py:549]: good call, copied from a composer test where the params aren't the same, but makes sense to add here.

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:34]: there are so many places to scrub the name "GPT" if thats what we want to do. but good callout

Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py:34]: labeling this as "outside the scope of this PR". @abhi-mosaic if we want to avoid the wrath of OpenAI/MS lawyers we'll eventually need to do this.

Comment (samhavens) [reply examples/llm/src/models/mosaic_gpt.py:34]: Hanlin said this was a long ways off. We had heard a rumor that PubMedGPT got renamed because of the GPT trademark thing, but it turned out to be because of the `PubMed` part. So we have plenty of lead time before we need to deal with this (if at all, we'll see if the courts say its enforceable)

**Comment by dakinggg** [comment]
Screen Shot 2023-03-21 at 5 09 45 PM
1bs training again

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:213]:
```suggestion
            past_key_values = [() for _ in range(self.config.n_layers)
```


[here](https://github.com/mosaicml/examples/pull/243/files#diff-cb5960418e22962455775ab1498957a896b97bdfbc69eab8444ee0779d224939R80) the type'd return type shows tuple
It doesn't matter if you keep the suggestion, just stay consistent.

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:275]: `if past_key_values is not None:` is it safe to assume `input_ids` is of shape B instead of Bx1? or BxSq?

File: examples/llm/tests/test_model.py
Comment (vchiley) [inline examples/llm/tests/test_model.py:1]: ngl these new tests are puzzling unless you're the one who wrote the tests...

**Review by vchiley — COMMENTED** [review]
Chonker PR!

File: examples/llm/src/models/configuration_mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/configuration_mosaic_gpt.py:138]: 👍

File: examples/llm/src/models/configuration_mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/configuration_mosaic_gpt.py:144]: 👍

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:158]: Well, it might. I guess I'm not sure if the forward nans cause issues and I didn't test. Should I a) remove this check without trying it b) leave the check and file a jira to test it out later c) hold this PR to test it

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:275]: Do you mean the input ids that are passed in here? I don't believe it is safe to assume that. Afaik this prepare function is the huggingface way to trim the inputs to a single token.

File: examples/llm/tests/test_model.py
Comment (dakinggg) [reply examples/llm/tests/test_model.py:1]: Ok yeah, that's fair. I will add some comments.

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:195]: Adding one isn't quite right I think.
If your input is of length 5, you'll start with
[0,1,2,3,4] pos ids. Then, if your padding mask is [0,0,1,1,1], the cumsum will be [1,2,2,2,2], which when subtracted off is [-1,-1,0,1,2], which is the correct pos ids for the unmasked positions, so you don't want to just add one. But the negatives will cause an error when passed to the embedding layer so need to be clamped to 0.

File: examples/llm/src/models/layers/attention.py
Comment (dakinggg) [reply examples/llm/src/models/layers/attention.py:312]: eh, I think I'll leave this one, since this class is pretty usable outside of MosaicGPT proper

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py:195]: (context: I'm just trying to get rid of the clamp...)

What about something like:
```
pos = torch.cumsum(attention_mask, dim=1)[:, past_position:]
```

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py:158]: d) convert check into warning, file a jira to test it out later

Comment (abhi-mosaic) [reply examples/llm/src/models/mosaic_gpt.py:158]: b) sounds good to me. We don't plan to train with left padding anytime soon, I think.

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py:195]: yeah I just don't like the ugly. I see my issue

File: examples/llm/src/models/mosaic_gpt.py
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:158]: I think I prefer the explicit error for now rather than a confusing failure. We can prioritize testing it out as needed, shouldn't take that long to try out and make sure convergence is preserved.

File: examples/llm/src/models/layers/attention.py
Comment (vchiley) [inline examples/llm/src/models/layers/attention.py:20]:
```suggestion
    if original_is_causal and num_query_tokens != num_key_tokens:
```

defaults to false if original_is_causal is false
avoids `raise NotImplementedError` if we don't need to check it.

File: examples/llm/src/models/mosaic_gpt.py
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:153]: do we want to point to a doc where these args are explained?
this'll help explain why we have the args, but then error out...

Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py:153]: let me add an explanatory comment in the code at least

**Review by vchiley — APPROVED** [review]
TY for the updates!

**Pull request merged**
